controllers/clusterurlmonitor/clusterurlmonitor_supplement.go

```diff
@@ -26,16 +28,31 @@ type BlackBoxExporter interface {
 	GetBlackBoxExporterNamespace() string
 }
 
+type ResourceComparer interface {
+	DeepEqual(x, y interface{}) bool
+}
+
 type ClusterUrlMonitorSupplement struct {
 	ClusterUrlMonitor v1alpha1.ClusterUrlMonitor
 	Client            client.Client
 	Log               logr.Logger
 	Ctx               context.Context
 	BlackBoxExporter  BlackBoxExporter
+	ResourceComparer  ResourceComparer
+}
+
+type ResourceComparerStruct struct{}
+
+func (_ ResourceComparerStruct) DeepEqual(x, y interface{}) bool {
+	return reflect.DeepEqual(x, y)
+}
+
+func DeepEqual(x, y interface{}) bool {
```

Comment:
We want to implement this to be able to mock DeepEqual for our unit testing.

Reply:
Do we really have to mock this function? It seems to be really flat, meaning no other in the tests unreachable dependencies are fetched. IMO this wouldn't require a mocking then, but just a few testcases where `x` and `y` are the same, and then where they are different. Although even this may not really be necessary, as we are just wrapping a 3rd party library `reflect` for the deep equal, which we can consider to be unit tested (when its a good one). In the unit tests we could skip by this further description of that `DeepEqual` in general for all function that we where it is used (it may come to use in quite a lot) and more focus on the input to ouput - behavior of our own function.

Reply:
Can we not even skip the wrapping and use the `reflect.DeepEqual` directly in the functions where we are using it?

Reply:
I am not sure if it was understood correctly but mocking DeepEqual is used on prometheusrule to ensure that the resource is equal to what we expect it to be.

Reply:
yup got it, I am just afraid that we are overdoing the mocking and make the unit tests too complicated. IMO we just have to mock unreachable external dependencies, which would be the kubernetes API. And for this cases we mocks we specify the return values for requests. The rest should be handled by our actual functions. 
For the case of the `EnsurePrometheusRuleResourceExists(ctx, Routemonitor)`, it would use our input the `RouteMonitor` and the return value of the `mockClient.Get`, that we specify in the beginning of the tests, and use the actual `reflect.DeepEqual`. But probably I don't understand fully how our current tests are put together.

Reply:
ok got it now, this mock allows us to avoid setting the value of the returned kubernetes resources that should be filled by the client mock. So I guess you can ignore my comments

Reply:
Can this comment be resolved now?

Reply:
yesyes